### PR TITLE
sql: label latency histograms with statement fingerprint

### PR DESCRIPTION
**Pull request by MattWhelan**
Previously, we collected a single histogram for all statements. While
this sort of metric may be useful in overall database operations,
application developers need more detail, to know how their code changes
impact their query latencies.

By switching to HistogramVec, we can now offer per-statement-fingerprint
latency metrics.

This feature is disabled by default. To know whether its safe to enable,
we also track a cardinality estimate of the unique statement
fingerprints that we see.

TREQ: https://cockroachlabs.atlassian.net/browse/TREQ-703

Epic: none

Release note (ops change):
Introduces a new `sql.exec.latency.detail` histogram metric. This metric
is labeled with its statement fingerprint. Enable this feature using the
`sql.stats.detailed_latency_metrics.enabled` application setting.

The `sql.query.unique.count` metric is a new count metric that estimates
the cardinality of the set of all statement fingerprints. For most
workloads, this ranges from dozens to hundreds. For workloads with over
a couple thousand fingerprints, we advise caution in enabling
`sql.stats.detailed_latency_metrics.enabled`.

Benchmark results:

```
# Baseline (master branch)
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
 1800.0s     6314.1  98.2%     31.2     21.0     52.4     88.1    209.7    570.4

# PR, setting disabled
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
 1800.0s     6315.6  98.2%     24.2     19.9     33.6     48.2    117.4    805.3

# PR, setting enabled
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
 1800.0s     6309.1  98.1%     24.2     19.9     33.6     48.2    113.2    671.1
```

### CODE REVIEW

**Review by dhartunian — COMMENTED** [review]
We have a workload that can simulate a high cardinality stream of fingerprints. You can run it via:

```
cockroach workload init sqlstats
cockroach workload run sqlstats
```

On a single node I observe that the unique counter is topping out at ~180, which is surprising, and the `_status/vars` output is 1.2G and takes a minute to write. I would expect in this case, the counter should at least signal to us that we shouldn't enable the feature.

```
~/go/src/github.com/cockroachdb/cockroach mwhelan/latencyFingerprint ≡
❯ http GET localhost:8080/_status/vars > metrics_lat4.out

~/go/src/github.com/cockroachdb/cockroach mwhelan/latencyFingerprint ≡ 1m
❯ ll | grep metrics_lat4.out
-rw-r--r--   1 davidh  staff   1.2G Nov 22 11:45 metrics_lat4.out

~/go/src/github.com/cockroachdb/cockroach mwhelan/latencyFingerprint ≡
❯ ag sql_query_unique_count metrics_lat4.out
6958827:# HELP sql_query_unique_count Cardinality estimate of the set of statement fingerprints
6958828:# TYPE sql_query_unique_count counter
6958829:sql_query_unique_count{node_id="1"} 186
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-:-OCJhNNL4Kk3rNtxeJdn:b9vt4ez)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @MattWhelan)

___
*[`-- commits` line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-OCJ_CpL-e6q826_S0SA:-OCJ_CpL-e6q826_S0SB:b-9ne3gb):*
nit: you need to use one of the pre-defined categories here. probably `ops change`.
___
*[`pkg/sql/executor_statement_metrics.go` line 317 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-OCJaa3L3LwZP3XadJxA:-OCJaa3L3LwZP3XadJxB:b-xjchl7) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ec0e872f1f180c8040c6591e667e85ecc79ec56/pkg/sql/executor_statement_metrics.go#L317)):*
> ```Go
> 			if shouldIncludeInLatencyMetrics {
> 				m.DistSQLExecLatency.Observe(labels, float64(runLatRaw.Nanoseconds()))
> 				m.DistSQLServiceLatency.RecordValue(svcLatRaw.Nanoseconds())
> ```

FYI there was an issue filed recently about the CPU impact of histogram recording in the prometheus library: https://github.com/cockroachdb/cockroach/issues/133306

Might be worth running a benchmark to see impact here, although if it's behind a setting there's more safety.
___
*[`pkg/sql/conn_executor.go` line 141 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-OCJ_uNT0q4giAc81WZv:-OCJ_uNT0q4giAc81WZw:binq5oo) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ec0e872f1f180c8040c6591e667e85ecc79ec56/pkg/sql/conn_executor.go#L141)):*
> ```Go
> var detailedLatencyMetrics = settings.RegisterBoolSetting(
> 	settings.ApplicationLevel,
> 	"server.detailed_latency_metrics.enabled",
> ```

This naming should be under the `sql.stats` namespace, not `server` which is more general purpose.
___
*[`pkg/sql/conn_executor.go` line 143 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-OCJa18ECoYbXu5z5F0H:-OCJa18ECoYbXu5z5F0I:b-atnur0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ec0e872f1f180c8040c6591e667e85ecc79ec56/pkg/sql/conn_executor.go#L143)):*
> ```Go
> 	"server.detailed_latency_metrics.enabled",
> 	"label latency metrics with the statement fingerprint. Workloads with tens of thousands of "+
> 		"distinct query fingerprints should leave this setting false.",
> ```

Consider adding validation that prevents enabling this when the cardinality estimate is too high. I'm on the fence about that since someone might still want to, for experiments.
___
*[`pkg/sql/conn_executor.go` line 528 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-OCJm7_F8UyZ8NTDFdHW:-OCJm7_F8UyZ8NTDFdHX:b-a5ft0e) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ec0e872f1f180c8040c6591e667e85ecc79ec56/pkg/sql/conn_executor.go#L528)):*
> ```Go
> 			// TODO(mrtracy): See HistogramWindowInterval in server/config.go for the 6x factor.
> 			DistSQLExecLatency: metric.NewExportedHistogramVec(
> 				getMetricMeta(MetaDistSQLExecLatency, internal),
> ```

This metric is output in both a standard an "internal" copy so the cardinality of the histogramvec itself is doubled. Consider not using it for `sql.exec.latency.internal`
___
*[`DEPS.bzl` line 780 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-OCJ_m5c-CIoBhSrCSCn:-OCJ_m5c-CIoBhSrCSCo:bwsujq2) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ec0e872f1f180c8040c6591e667e85ecc79ec56/DEPS.bzl#L780)):*
> ```Starlark
>         build_file_proto_mode = "disable_global",
>         importpath = "github.com/axiomhq/hyperloglog",
>         # TODO: mirror this repo (to fix, run `./dev generate bazel --mirror`)
> ```

FYI you'll need to do this prior to merge

**Review by MattWhelan — COMMENTED** [review]
That test workload is useful, thanks. I see similar results. After about a minute, I see about a half-million distinct insert fingerprints, but the cardinality estimate is only 145. 

Which is super odd. My unit tests for the UniqueCount metric look fine. If you have any ideas what I did wrong here, I'm all ears. I'll be digging into this issue. We can't really have this feature without a uniquecount metric, and hyperloglog isn't exactly an unproven technology :)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-:-OCKo22r2bobb0Ab9Mdf:b-vwicjh)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

___
*[`DEPS.bzl` line 780 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-OCJ_m5c-CIoBhSrCSCn:-OCKoyKa25sbl5RaofmU:b-1m7na) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ec0e872f1f180c8040c6591e667e85ecc79ec56/DEPS.bzl#L780)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

FYI you'll need to do this prior to merge
</blockquote></details>

Yeah. I need to get access to that bucket first.
___
*[`pkg/sql/conn_executor.go` line 141 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-OCJ_uNT0q4giAc81WZv:-OCKq_6X77Cgdbcd8cP7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ec0e872f1f180c8040c6591e667e85ecc79ec56/pkg/sql/conn_executor.go#L141)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

This naming should be under the `sql.stats` namespace, not `server` which is more general purpose.
</blockquote></details>

Done.
___
*[`pkg/sql/conn_executor.go` line 143 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-OCJa18ECoYbXu5z5F0H:-OCKqeYk1PrQ8IocDY9X:b-6cwura) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ec0e872f1f180c8040c6591e667e85ecc79ec56/pkg/sql/conn_executor.go#L143)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Consider adding validation that prevents enabling this when the cardinality estimate is too high. I'm on the fence about that since someone might still want to, for experiments.
</blockquote></details>

Since it defaults to off, and the real-world statement cardinality isn't generally problematic, I'm inclined to let users have it if they want it.
___
*[`pkg/sql/conn_executor.go` line 528 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-OCJm7_F8UyZ8NTDFdHW:-OCKrCYCC0sgqXp-uH3u:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ec0e872f1f180c8040c6591e667e85ecc79ec56/pkg/sql/conn_executor.go#L528)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

This metric is output in both a standard an "internal" copy so the cardinality of the histogramvec itself is doubled. Consider not using it for `sql.exec.latency.internal`
</blockquote></details>

Done
___
*[`pkg/sql/executor_statement_metrics.go` line 317 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-OCJaa3L3LwZP3XadJxA:-OCKsNbq0YnbHBa73IYB:bs2ahqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ec0e872f1f180c8040c6591e667e85ecc79ec56/pkg/sql/executor_statement_metrics.go#L317)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

FYI there was an issue filed recently about the CPU impact of histogram recording in the prometheus library: https://github.com/cockroachdb/cockroach/issues/133306

Might be worth running a benchmark to see impact here, although if it's behind a setting there's more safety.
</blockquote></details>

I see that a more recent version of the prometheus client library also has performance improvements related to allocations in MetricVec. 

Prometheus has kind of an odd approach to histograms: each bucket is cumulative, so you have to increment the count for the bucket the event falls into, and all the higher-cutoff buckets. I truly don't understand why you'd choose to do that at recording time, but with a lot of buckets, it's a lot of writes. 

There's a microbenchmark in here; the performance impact on even relatively high cardinality workloads is pretty minor: dozens of nanoseconds in the metric recording method. I'm planning to do a whole-system benchmark as well before I mark the PR ready.

**Review by MattWhelan — COMMENTED** [review]
I think I figured out the cardinality issue. I had ignored the `internal` flag in the metric setup code, which apparently resulted in the internal metric being exported, but not the main one.

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-:-OCL0ud05io1rpE3w-bf:bujskv7)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

**Review by dhartunian — APPROVED** [review]
LGTM! Thanks for your work on this.

One thing on my mind for customers on this feature is that I can imagine someone turning this on and then getting a cardinality explosion by accident when their workflow changes many months later. What do you think of applying a cap to the cardinality?

> Reviewed 6 of 9 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-:-ODMN9d08rIpbsl8BxhS:b-iir8tm)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, and @xinhaoz)

___
*[`pkg/sql/conn_executor.go` line 143 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-OCJa18ECoYbXu5z5F0H:-ODIt_HyEaLTnVcV2U4w:bb73djm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ec0e872f1f180c8040c6591e667e85ecc79ec56/pkg/sql/conn_executor.go#L143)):*
<details><summary><i>Previously, MattWhelan (Matt Whelan) wrote…</i></summary><blockquote>

Since it defaults to off, and the real-world statement cardinality isn't generally problematic, I'm inclined to let users have it if they want it.
</blockquote></details>

SGTM

**Review by MattWhelan — COMMENTED** [review]
I honestly think it's unlikely to cause instability in real workloads. It could create resource/performance issues though. 

I think the best place to put such a limit is in the metric package itself. The `vector` type, perhaps, could enforce a cardinality limit. It's aware of the current cardinality. Integrating it there would also solve the issue for any future MetricVec style metrics driven by user data. 

Maybe make `vector.recordLabels` fallible, and skip the call to `Observe` if it fails?

Would we want a hard-coded limit, or a cluster setting?

> Reviewed 2 of 13 files at r1, 2 of 3 files at r2, 9 of 9 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-:-ODNDLI25JWtnJGbOb5V:b3l3fbt)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @kyle-a-wong, and @xinhaoz)

**Review by kyle-a-wong — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135924)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @MattWhelan, and @xinhaoz)

___
*[`pkg/sql/conn_executor.go` line 525 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-ODN4HEe0aCTJrtvdzIM:-ODN4HEe0aCTJrtvdzIN:biwi8zt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac54648cf7a05bca75517b97baf8e362f60e0891/pkg/sql/conn_executor.go#L525)):*
> ```Go
> 	}
> 
> 	distSqlExecLatency := metric.NewExportedHistogramVec(
> ```

As per the doc string for `NewExportedHistogramVec`, this metric type doesn't get persisted to tsdb (even if the cluster setting is disabled). Is this what you want?:
```
// NewExportedHistogramVec creates a new HistogramVec containing labeled counters to
// be exported to an external collector; the contained histograms are not
// aggregated or persisted to the tsdb
```

**Review by MattWhelan — COMMENTED** [review]
Ok, I pushed up a cardinality limit. What do you think of it?

> Reviewed 1 of 9 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-:-ODNmr5T4jeJySkaBs6h:bo6sz39)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, @kyle-a-wong, and @xinhaoz)

___
*[`pkg/sql/conn_executor.go` line 525 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-ODN4HEe0aCTJrtvdzIM:-ODNn03CDR1uFc_sq4g_:bk9ag0o) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac54648cf7a05bca75517b97baf8e362f60e0891/pkg/sql/conn_executor.go#L525)):*
<details><summary><i>Previously, kyle-a-wong (Kyle Wong) wrote…</i></summary><blockquote>

As per the doc string for `NewExportedHistogramVec`, this metric type doesn't get persisted to tsdb (even if the cluster setting is disabled). Is this what you want?:
```
// NewExportedHistogramVec creates a new HistogramVec containing labeled counters to
// be exported to an external collector; the contained histograms are not
// aggregated or persisted to the tsdb
```
</blockquote></details>

Hmm. Help me out here: what will that impact?

**Review by kyle-a-wong — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135924)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, @MattWhelan, and @xinhaoz)

___
*[`pkg/sql/conn_executor.go` line 525 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-ODN4HEe0aCTJrtvdzIM:-ODRDj9B9QPxV-4f8yEc:bwvj256) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac54648cf7a05bca75517b97baf8e362f60e0891/pkg/sql/conn_executor.go#L525)):*
<details><summary><i>Previously, MattWhelan (Matt Whelan) wrote…</i></summary><blockquote>

Hmm. Help me out here: what will that impact?
</blockquote></details>

So the other histogram metric types in crdb (created via `metric.NewHistogram` pre-compute / aggregate various histogram metrics and store them into tsdb. This includes `<name>-count`, `name-sum`, and the suffix listed here: https://github.com/cockroachdb/cockroach/blob/64c9018ac88bcc2f16b0f7ad0c73bb4e60b770fd/pkg/util/metric/metric.go#L1062. tsdb doesnt actually store the raw histogram data. All of the metrics in the db-console metrics dashboard and custom chart pages are powered using metrics from tsdb. One thing to note about tsdb is that it doesn't really work with labeled metrics, tsdb stores the metrics with the key as the metric name itself and doesn't use the labels of the metrics.

The type you are using, and the other vector types in metrics.go, are intended to be use for metrics where we want to include label information, but this mean's that we can't store them in tsdb. These metrics will only be exported as prometheus metrics and are only accessible via `/_status/vars`. As a result, we've named the constructors with the pattern: `NewExported*` to try to help indicate this as well. It should also be noted that this `HistogramVec` type doesn't do any pre-computation or aggregation, it just puts the raw prometheus histogram metrics into `/_status/vars` (meaning `<name>_count`,  `<name>_sum`, and ` `<name>_bucket` metrics)

TLDR: 
 - The `sql.distsql.exec.latency` metric will no longer be visible in db-console's metrics page
 - Pre computation and aggregated metrics will no longer exist in scraped metrics

**Review by MattWhelan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135924)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, @kyle-a-wong, and @xinhaoz)

___
*[`pkg/sql/conn_executor.go` line 525 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-ODN4HEe0aCTJrtvdzIM:-ODRxh5j0r3fkpxInzDr:b-dj4j5z) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac54648cf7a05bca75517b97baf8e362f60e0891/pkg/sql/conn_executor.go#L525)):*
<details><summary><i>Previously, kyle-a-wong (Kyle Wong) wrote…</i></summary><blockquote>

So the other histogram metric types in crdb (created via `metric.NewHistogram` pre-compute / aggregate various histogram metrics and store them into tsdb. This includes `<name>-count`, `name-sum`, and the suffix listed here: https://github.com/cockroachdb/cockroach/blob/64c9018ac88bcc2f16b0f7ad0c73bb4e60b770fd/pkg/util/metric/metric.go#L1062. tsdb doesnt actually store the raw histogram data. All of the metrics in the db-console metrics dashboard and custom chart pages are powered using metrics from tsdb. One thing to note about tsdb is that it doesn't really work with labeled metrics, tsdb stores the metrics with the key as the metric name itself and doesn't use the labels of the metrics.

The type you are using, and the other vector types in metrics.go, are intended to be use for metrics where we want to include label information, but this mean's that we can't store them in tsdb. These metrics will only be exported as prometheus metrics and are only accessible via `/_status/vars`. As a result, we've named the constructors with the pattern: `NewExported*` to try to help indicate this as well. It should also be noted that this `HistogramVec` type doesn't do any pre-computation or aggregation, it just puts the raw prometheus histogram metrics into `/_status/vars` (meaning `<name>_count`,  `<name>_sum`, and ` `<name>_bucket` metrics)

TLDR: 
 - The `sql.distsql.exec.latency` metric will no longer be visible in db-console's metrics page
 - Pre computation and aggregated metrics will no longer exist in scraped metrics
</blockquote></details>

Ok, that makes sense. And removing this metric from the console seems like a problem. 

So it looks like we actually store most histogram types twice: once as "windowed" delta histograms, and once as cumulative prometheus-style histograms. But not the newer vec ones: they only support the cumulative format. Which is fine, since the point is the labels, and those are, at best, lost in the internal TSDB. 

The obvious thing to do would be to just make these metrics new: give them a different name, and leave the old ones in place. There's a trivial cost to that, but it won't break things. Clearly, since TSDB doesn't do labels, we can't actually have this data there in a meaningful way, so there's no real downside to that implementation. 

Does that seem like a reasonable way forward?

**Review by MattWhelan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135924)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, @kyle-a-wong, and @xinhaoz)

___
*[`pkg/sql/conn_executor.go` line 525 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-ODN4HEe0aCTJrtvdzIM:-ODStkWQCAkJmqsmhRMM:b4j2pyy) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac54648cf7a05bca75517b97baf8e362f60e0891/pkg/sql/conn_executor.go#L525)):*
<details><summary><i>Previously, MattWhelan (Matt Whelan) wrote…</i></summary><blockquote>

Ok, that makes sense. And removing this metric from the console seems like a problem. 

So it looks like we actually store most histogram types twice: once as "windowed" delta histograms, and once as cumulative prometheus-style histograms. But not the newer vec ones: they only support the cumulative format. Which is fine, since the point is the labels, and those are, at best, lost in the internal TSDB. 

The obvious thing to do would be to just make these metrics new: give them a different name, and leave the old ones in place. There's a trivial cost to that, but it won't break things. Clearly, since TSDB doesn't do labels, we can't actually have this data there in a meaningful way, so there's no real downside to that implementation. 

Does that seem like a reasonable way forward?
</blockquote></details>

Ok, pivoted to a new metric name for the new histograms to avoid regressions in the existing observability space.

**Review by kyle-a-wong — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135924)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, @MattWhelan, and @xinhaoz)

___
*[`pkg/sql/conn_executor.go` line 525 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-ODN4HEe0aCTJrtvdzIM:-ODSyYoB-2FMJjmxuMYB:bwrzip6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac54648cf7a05bca75517b97baf8e362f60e0891/pkg/sql/conn_executor.go#L525)):*
<details><summary><i>Previously, MattWhelan (Matt Whelan) wrote…</i></summary><blockquote>

Ok, pivoted to a new metric name for the new histograms to avoid regressions in the existing observability space.
</blockquote></details>

Sounds good to me. I'll give the PR another look through

**Review by kyle-a-wong — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135924)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, @MattWhelan, and @xinhaoz)

___
*[`pkg/sql/executor_statement_metrics.go` line 312 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-ODgIzRv0rqky3bQj4Wq:-ODgIzRv0rqky3bQj4Wr:b-n33hnb) ([raw file](https://github.com/cockroachdb/cockroach/blob/d55aa02bc71a99ad71239a9473a5e2e390078e27/pkg/sql/executor_statement_metrics.go#L312)):*
> ```Go
> 		if detailedLatencyMetrics.Get(&ex.server.cfg.Settings.SV) {
> 			labels = map[string]string{
> 				detailedLatencyMetricLabel: stmt.StmtNoConstants,
> ```

Its a little worrying to me we are using `stmt.StmtNoConstants` as the label value here. This can potentially be a very long string. It may be better to do something like truncate this to be 50-100 characters, and maybe include the statement fingerprint id or something. What do you think?



_Code quote:_
```Go
stmt.StmtNoConstants
```

**Review by MattWhelan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135924)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, @kyle-a-wong, and @xinhaoz)

___
*[`pkg/sql/executor_statement_metrics.go` line 312 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-ODgIzRv0rqky3bQj4Wq:-ODhDtZz1PQiMPqOjPu_:b-y5welh) ([raw file](https://github.com/cockroachdb/cockroach/blob/d55aa02bc71a99ad71239a9473a5e2e390078e27/pkg/sql/executor_statement_metrics.go#L312)):*
<details><summary><i>Previously, kyle-a-wong (Kyle Wong) wrote…</i></summary><blockquote>

Its a little worrying to me we are using `stmt.StmtNoConstants` as the label value here. This can potentially be a very long string. It may be better to do something like truncate this to be 50-100 characters, and maybe include the statement fingerprint id or something. What do you think?
</blockquote></details>

I considered that approach, but after researching the total size of unique fingerprints in real workloads, I decided against it. The actual size of this set of strings is small in all but the most extreme workloads. There's conversation about that here: https://docs.google.com/document/d/1skM28AnwJ-IZ_LQLhc3er_K76wHTJWFu-Gnj-zTcljA/edit?tab=t.0#heading=h.9ndo02nsmu

The purpose of the feature is to allow users to notice changes in latency and trace them to changes in their queries. For long queries, it's entirely possible for the unique parts to be at the end of the query. Providing a numeric hash isn't going to help trace a latency change back to a code change.

**Review by kyle-a-wong — APPROVED** [review]
The commit message is out of date now with the last revision you made, can you update it to reflect the introduction of the new metric, instead of change the existing metric. Also, can you include the benchmark results in the commit message / PR body?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-:-ODmEJIhD859HMbBMyZS:b-tp34b6)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, @MattWhelan, and @xinhaoz)

___
*[`pkg/sql/executor_statement_metrics.go` line 312 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-ODgIzRv0rqky3bQj4Wq:-ODm2BPK7SlOqcV0aHTU:b9qk6bj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d55aa02bc71a99ad71239a9473a5e2e390078e27/pkg/sql/executor_statement_metrics.go#L312)):*
<details><summary><i>Previously, MattWhelan (Matt Whelan) wrote…</i></summary><blockquote>

I considered that approach, but after researching the total size of unique fingerprints in real workloads, I decided against it. The actual size of this set of strings is small in all but the most extreme workloads. There's conversation about that here: https://docs.google.com/document/d/1skM28AnwJ-IZ_LQLhc3er_K76wHTJWFu-Gnj-zTcljA/edit?tab=t.0#heading=h.9ndo02nsmu

The purpose of the feature is to allow users to notice changes in latency and trace them to changes in their queries. For long queries, it's entirely possible for the unique parts to be at the end of the query. Providing a numeric hash isn't going to help trace a latency change back to a code change.
</blockquote></details>

I see. Thanks for linking the document. I think one thing about longer label values that worried me / worries me was potential for an error to occur, similar to an error mentioned in this TREQ(https://cockroachlabs.atlassian.net/browse/TREQ-613) for help text that is too long. After some research, it doesn't seem like a similar error exists for label values.

**Review by MattWhelan — COMMENTED** [review]
I've already updated the commit message. Is there something I missed?

I'll summarize the benchmark results. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-:-ODqjwNu4AK08zIMzPvm:btp8d9g)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, @kyle-a-wong, and @xinhaoz)

___
*[`pkg/sql/executor_statement_metrics.go` line 312 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-ODgIzRv0rqky3bQj4Wq:-ODqiwxL2woTHZ2TJtco:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d55aa02bc71a99ad71239a9473a5e2e390078e27/pkg/sql/executor_statement_metrics.go#L312)):*
<details><summary><i>Previously, kyle-a-wong (Kyle Wong) wrote…</i></summary><blockquote>

I see. Thanks for linking the document. I think one thing about longer label values that worried me / worries me was potential for an error to occur, similar to an error mentioned in this TREQ(https://cockroachlabs.atlassian.net/browse/TREQ-613) for help text that is too long. After some research, it doesn't seem like a similar error exists for label values.
</blockquote></details>

Done.

**Review by kyle-a-wong — COMMENTED** [review]
Ohh okay, I'm not sure where the root message of this top level discussion comes from, but It seems like it doesn't get updated with the commit message. Maybe its the PR description or something. Anyway, i've left a comment on the commit message itself.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-:-ODqkr5Q-yoe3j68YL5v:bx6hitc)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, @MattWhelan, and @xinhaoz)

___
*[`-- commits` line 10 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-ODqlVUg4HDEOIoh16iW:-ODqlVUg4HDEOIoh16iX:b-2x80n6):*
nit: Now that the implementation has changed, the description is not correct. We didn't switch to a HistogramVec, we created a new metric, which uses a HistogramVec, that provides more detailed metrics.



_Code quote:_
```
  Previously, we collected a single histogram for all statements. While
  this sort of metric may be useful in overall database operations,
  application developers need more detail, to know how their code changes
  impact their query latencies.

  By switching to HistogramVec, we can now offer per-statement-fingerprint
  latency metrics.
```

**Review by kyle-a-wong — APPROVED** [review]
Oh i see that it updated, so it looks like it is synced with the PR message. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135924#-:-ODqmj6x2tNdn-lMwxOg:byema8u)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, @MattWhelan, and @xinhaoz)
